go/store/nbs/store.go

```diff
@@ -390,6 +393,66 @@ func fromManifestAppendixOptionNewContents(upstream manifestContents, appendixSp
 	}
 }
 
+// OverwriteManifestAndClose overwrites the existing manifest with the given root hash, tableFiles,
+// and appendixTableFiles. After writing the manifest, it closes the store since it is now out of date.
+// It returns the updated manifest contents.
+func (nbs *NomsBlockStore) OverwriteManifestAndClose(ctx context.Context, root hash.Hash, tableFiles map[hash.Hash]uint32, appendixTableFiles map[hash.Hash]uint32) (mi ManifestInfo, err error) {
+	nbs.mm.LockForUpdate()
+	defer func() {
+		unlockErr := nbs.mm.UnlockForUpdate()
+
+		if err == nil {
+			err = unlockErr
+		}
+	}()
+
+	nbs.mu.Lock()
+	defer nbs.mu.Unlock()
+
+	var originalLock addr
+	{
+		_, c, err := nbs.mm.Fetch(ctx, nbs.stats)
+		if err != nil {
+			return manifestContents{}, err
+		}
+		originalLock = c.lock
+	}
```

Comment:
This seems wrong to me. If we are overwriting the manifest, not appending to it, then we should concurrency control on `nbs.upstream.lock`. We don't need to fetch the current manifest from manifest manager; we need to use our upstream's view to know if we failed the optimistic write.